execute/metrics/prom.go

```diff
@@ -76,114 +50,71 @@ func NewPromReporter(lggr logger.Logger, selector cciptypes.ChainSelector) (*Pro
-		counters[tokenDataLabel] += len(report.OffchainTokenData)
-
-		p.lggr.Debugw("Execute plugin reporting outcome",
-			"state", state,
-			"sourceChainSelector", chainSelector,
-			"destChainSelector", p.chainID,
-			"messagesCount", len(report.Messages),
-			"tokenDataCount", len(report.OffchainTokenData),
-		)
-	}
+func (p *PromReporter) TrackOutcome(outcome exectypes.Outcome, state exectypes.PluginState) {
+	p.trackOutputStats(&outcome, state, plugincommon.OutcomeMethod)
 
-	for key, count := range counters {
-		p.outcomeDetailsCounter.
-			WithLabelValues(p.chainID, state, key).
-			Add(float64(count))
+	for _, cr := range outcome.CommitReports {
+		sourceChainSelector := cr.SourceChain
+		maxSeqNr := pickHighestSeqNrInMessages(cr.Messages)
+		p.trackMaxSequenceNumber(sourceChainSelector, maxSeqNr, plugincommon.OutcomeMethod)
 	}
 }
 
-func (p *PromReporter) trackCommitReports(commits exectypes.CommitObservations, state string) {
-	commitReportsCount := 0
-	for _, commit := range commits {
-		commitReportsCount += len(commit)
+func (p *PromReporter) trackMaxSequenceNumber(
+	sourceChainSelector cciptypes.ChainSelector,
+	maxSeqNr int,
+	method plugincommon.MethodType,
+) {
+	if maxSeqNr == 0 {
+		return
 	}
 
-	p.observationDetailsCounter.
-		WithLabelValues(p.chainID, state, commitReportsLabel).
-		Add(float64(commitReportsCount))
+	p.sequenceNumbers.
+		WithLabelValues(p.chainID, sourceChainSelector.String(), method).
+		Set(float64(maxSeqNr))
 }
 
-func (p *PromReporter) trackMessages(messages exectypes.MessageObservations, state string) {
-	messagesCount := 0
-	for _, chainMessages := range messages {
-		messagesCount += len(chainMessages)
+func (p *PromReporter) trackOutputStats(
+	output plugintypes.Trackable,
+	state exectypes.PluginState,
+	method plugincommon.MethodType,
+) {
+	stringState := string(state)
+	for key, val := range output.Stats() {
+		p.outputDetailsCounter.
+			WithLabelValues(p.chainID, method, stringState, key).
+			Add(float64(val))
 	}
-
-	p.observationDetailsCounter.
-		WithLabelValues(p.chainID, state, messagesLabel).
-		Add(float64(messagesCount))
 }
 
-func (p *PromReporter) trackTokenData(tokens exectypes.TokenDataObservations, state string) {
-	tokenCounters := map[string]int{
-		tokenStateReady:   0,
-		tokenStateWaiting: 0,
-	}
-
-	for _, chainTokens := range tokens {
-		for _, tokenData := range chainTokens {
-			for _, token := range tokenData.TokenData {
-				counterKey := tokenStateWaiting
-				if token.IsReady() {
-					counterKey = tokenStateReady
-				}
-				tokenCounters[counterKey]++
-			}
-		}
-	}
-
-	for key, count := range tokenCounters {
-		p.tokenDataReadinessCounter.
-			WithLabelValues(p.chainID, state, key).
-			Add(float64(count))
+func pickHighestSeqNrInMessages(messages []cciptypes.Message) int {
+	seqNrs := make([]cciptypes.SeqNum, len(messages))
+	for i, m := range messages {
+		seqNrs[i] = m.Header.SequenceNumber
 	}
+	return pickHighestSeqNr(seqNrs)
```

Comment:
Maybe https://pkg.go.dev/slices#Max

Reply:
Shoot, I've tried that, but it panics when slice is empty